// swapi/dist/src/decorators/http.methods.js
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../types");
const helpers_1 = require("../helpers");
const storage_1 = require("../storage");
const normalize_path_1 = require("../helpers/normalize.path");
function Get(path, description = '') {
    return BasicHttpMethodDecorator(path, types_1.HttpMethods.GET, description);
}
exports.Get = Get;
function Post(path, description = '') {
    return BasicHttpMethodDecorator(path, types_1.HttpMethods.POST, description);
}
exports.Post = Post;
function Put(path, description = '') {
    return BasicHttpMethodDecorator(path, types_1.HttpMethods.PUT, description);
}
exports.Put = Put;
function Patch(path, description = '') {
    return BasicHttpMethodDecorator(path, types_1.HttpMethods.PATCH, description);
}
exports.Patch = Patch;
function Delete(path, description = '') {
    return BasicHttpMethodDecorator(path, types_1.HttpMethods.DELETE, description);
}
exports.Delete = Delete;
function BasicHttpMethodDecorator(path, method, description = '') {
    path = normalize_path_1.normalizePath(path);
    const urlParams = helpers_1.pullOutParamsFromUrl(path);
    return (target, endpointName, descriptor) => {
        const nodeName = target.constructor.name;
        const endpoint = {
            name: endpointName,
            path,
            method,
            description,
            urlParams
        };
        storage_1.NodeStorage.getInstance().upsertEndpoint(nodeName, endpoint.name, endpoint);
    };
}
